### PR TITLE
Read sources and kerning

### LINKED ISSUE #48: Read font sources
As a follow up it would be good to be able to read at least the basic master information and translate it to fontra font source which include information about name, location, italic angle, vertical metrics.

### LINKED ISSUE #51: Read kerning info
This depends on:
- googlefonts/fontra/pull/1500
- #48

We need to extract:
- kerning groups: convert glyph-level kern "keys" to a UFO style group `dict[str, list[str]]`
- kerning values

We will also need to extract vertical kerning + groups. This could be done as a separate task, but it is equally urgent.

### PR DESCRIPTION
**Pull request by ollimeier**
Fixes #48 
Fixes #51 

This PR reads GlyphsApp masters and translates it into Fontra font sources + reads kerning

### CODE REVIEW

File: tests/test_backend.py [hunk 1]
```diff
@@ -113,3 +113,5 @@ async def test_glyphRead(testFont, referenceFont, glyphName):
 
     referenceGlyph = await referenceFont.getGlyph(glyphName)
     assert referenceGlyph == glyph
+
+    assert await testFont.getSources() == await referenceFont.getSources()
```
Comment (justvanrossum) [inline tests/test_backend.py:117]: Please put this in its own test func. This one is called `test_readGlyph()`, and should keep focusing on that.

File: src/fontra_glyphs/backend.py [hunk 2]
```diff
@@ -457,3 +469,74 @@ def fixSourceLocations(sources, smartAxisNames):
...
+
+
+def gsToFontraZone(gsVerticalMetricsValue, gsAlignmentZones):
+    for gsZone in gsAlignmentZones:
+        if gsZone.position == gsVerticalMetricsValue:
+            return gsZone.size
+    return 0
+
+
+def gsVerticalMetricToFontraLineMetric(value, zone):
```
Comment (justvanrossum) [inline src/fontra_glyphs/backend.py:499]: This function seems redundant. Compare:
```python
    gsVerticalMetricToFontraLineMetric(A, B)
```
with
```python
    LineMetric(value=A, zone=B)
```

File: src/fontra_glyphs/backend.py [hunk 3]
```diff
@@ -457,3 +469,74 @@ def fixSourceLocations(sources, smartAxisNames):
...
+
+
+def gsMastersToFontraFontSources(gsFont):
+    sources = {}
+    for gsMaster in gsFont.masters:
+        sources[gsMaster.id] = FontSource(
+            name=gsMaster.name,
+            italicAngle=gsMaster.italicAngle,
+            location={},
+            lineMetricsVerticalLayout=gsVerticalMetricsToFontraLineMetricsVertical(
```
Comment (justvanrossum) [inline src/fontra_glyphs/backend.py:481]: I'm guessing you mean `lineMetricsHorizontalLayout`.

`lineMetricsHorizontalLayout` are the line metrics for _horizontal_ (LTR or RTL) writing / text layout. `lineMetricsVerticalLayout` are the line metrics for _vertical_ (TTB/BTT) writing / text layout.

Comment (justvanrossum) [reply src/fontra_glyphs/backend.py:481]: Please also test with a "real" .glyphs file beyond unit testing, to check whether you actually see the values displayed.

File: tests/test_backend.py (hunk 1 above)
Comment (ollimeier) [reply tests/test_backend.py:119]: done.

File: src/fontra_glyphs/backend.py (hunk 2 above)
Comment (ollimeier) [reply src/fontra_glyphs/backend.py]: done.

File: src/fontra_glyphs/backend.py (hunk 3 above)
Comment (ollimeier) [reply src/fontra_glyphs/backend.py]: Thanks for the hint. I missed to work on `location={},` why it fails to display the sources within Fontra.

Comment (ollimeier) [reply src/fontra_glyphs/backend.py]: done and works:
Screenshot 2024-06-28 at 12 43 23

File: src/fontra_glyphs/backend.py [hunk 4]
```diff
@@ -489,49 +494,37 @@ def fixSourceLocations(sources, smartAxisNames):
...
+    return newPrefix + name[oldPrefixLength:] if name.startswith(oldPrefix) else name
 
 
 def getNormalizedKerningDict(gsFont, gsMasterID, valueDicts):
     kernDict = gsFont.kerning[gsMasterID]
-    for leftKernID in kernDict.keys():
-        leftKey, fontraLeftKey = getKerningNameFromID(gsFont, leftKernID)
-        if not leftKey:
-            continue
+    for left, values in kernDict.items():
```
Comment (justvanrossum) [inline src/fontra_glyphs/backend.py:504]: Suggest to rename `values` to `rightDict`

**Comment by justvanrossum** [comment]
This is more or less ready. It would be nice to have test data for vertical kerning: I tested it manually with a non-sharable .glyphs file with vertical kerning, but ideally the unit tests should cover it, too.

File: src/fontra_glyphs/backend.py [hunk 5]
```diff
@@ -457,3 +524,110 @@ def fixSourceLocations(sources, smartAxisNames):
...
+    fontraPrefix2 = FONTRA_KERN_GROUP_PREFIXES[side2]
+
+    groups = dict(groupsBySide[side1] | groupsBySide[side2])
+
+    sourceIdentifiers = []
+    valueDicts: dict[str, dict[str, dict]] = defaultdict(lambda: defaultdict(dict))
+
+    for gsMaster in gsFont.masters:
+        kernDict = getattr(gsFont, kerningAttr).get(gsMaster.id)
+        if kernDict is None:
```
Comment (justvanrossum) [inline src/fontra_glyphs/backend.py:548]: I have run into the situation where some masters have kerning, but the default (neutral) master doesn't. 

This is problematic for interpolation later on (can't interpolate without the "base master").

I can fix it by not skipping masters without kerning (by ensuring kernDict is `{}` in those cases), but I think a nicer solution would be if we _do_ skip kerning-less masters, _except_ when it's the default master. Unfortunately it is not clear to me how to get the "base master" from a GSFont.

Comment (justvanrossum) [reply src/fontra_glyphs/backend.py:548]: Solution: `get_regular_master()`.

**Pull request merged**